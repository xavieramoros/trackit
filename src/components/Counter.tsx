import React, { useState, useRef } from 'react';
import { View, Text, StyleSheet } from 'react-native';

import { addSeconds } from 'date-fns'

import { Button, Icon } from '@components/Themed';
import { convertSecondsToTime } from '@utils/time';

type CounterState = 'initial' | 'playing' | 'paused';

type callbackPayload = {
  state: CounterState,
  count: number
}

type CounterProps = {
  onPlay?: ({ state } : { state: CounterState}) => void,
  onPause?: ({ count, state} : callbackPayload) => void,
  onStop?: ({ count, state} : callbackPayload) => void,
  initialCount?: number
}

const Counter = ({ onStop = () => {}, onPlay = () => {}, onPause = () => {}, initialCount = 0}: CounterProps) => {
  const [count, setCount] = useState(initialCount);
  const [state, setState] = useState<CounterState>('initial');
  const intervalRef = useRef<NodeJS.Timeout>();

  const startCount = () => {
    intervalRef.current = setInterval(() => {
      setCount((prevCount: number) => prevCount + 1);
    }, 1000);

    setState('playing');
    onPlay({ state: 'playing'});
  };

  const pauseCount = () => {
    clearInterval(intervalRef.current);
    setState('paused');
    onPause({ count, state: 'paused' })
  };

  const stopCount = () => {
    clearInterval(intervalRef.current);
    onStop({ count, state: 'initial' });
    setCount(0);
    setState('initial');
  };

  const { hours, minutes, seconds } =  convertSecondsToTime(count)
  return (
    <View style={styles.container}>
      <View style={styles.countContainer}>
        <Text style={styles.countText}>
          { `${hours}:${minutes}:${seconds}`}
        </Text>
      </View>
      <View style={styles.buttonContainer}>
        {state === 'initial' && (
          <Button
            icon={<Icon name="play-circle" type="font-awesome" color="white" />}
            onPress={startCount}
          />
        )}
        {state === 'playing' && (
          <>
            <Button
              icon={<Icon name="pause-circle" type="font-awesome" color="white" />}
              onPress={pauseCount}
            />
            <Button
                containerStyle={styles.marginLeft}
                icon={<Icon name="stop-circle" type="font-awesome" color="white" />}
                onPress={stopCount}
            />
          </>
        )}
        {state === 'paused' && (
          <>
            <Button
              icon={<Icon name="play-circle" type="font-awesome" color="white" />}
              onPress={startCount}
            />
            <Button
              containerStyle={styles.marginLeft}
              icon={<Icon name="stop-circle" type="font-awesome" color="white" />}
              onPress={stopCount}
            />
          </>
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  countContainer: {
    marginBottom: 20,
  },
  countText: {
    fontSize: 40,
  },
  buttonContainer: {
    flexDirection: 'row',
  },
  marginLeft: {
    marginLeft: 8
  }
});

export default Counter;
