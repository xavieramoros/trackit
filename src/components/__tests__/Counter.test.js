import renderer from "react-test-renderer";
import { render, fireEvent, act } from '@testing-library/react-native';

import Counter from '../Counter';

let mockedProps;
const initialTimestamp = Date.now() - 5000;

describe(`Counter `, () => {
  it('renders the initial state correctly', () => {
    const { getByText, queryByText, queryByTestId } = render(
      <Counter />
    );

    expect(queryByTestId('play')).not.toBeNull();
    expect(queryByText(/00:00:00/i)).not.toBeNull();
    expect(queryByTestId('pause')).toBeNull();
    expect(queryByTestId('stop')).toBeNull();
  });

  it('renders the play button and triggers onPlay callback when pressed', () => {
    const onPlayMock = jest.fn();
    const { getByTestId } = render(<Counter onPlay={onPlayMock} />);

    act(() => {
      fireEvent.press(getByTestId('play'));
      jest.advanceTimersByTime(1000);
    });


    expect(onPlayMock).toHaveBeenCalledTimes(1);
    expect(getByTestId('pause')).not.toBeNull();
    expect(getByTestId('stop')).not.toBeNull();
  });

  it('pauses the counter when the pause button is pressed', async () => {
    const onPauseMock = jest.fn();
    const { getByTestId } = render(
      <Counter onPause={onPauseMock} initialTimestamp={initialTimestamp} initialState="playing" />
    );

    act(() => {
      fireEvent.press(getByTestId('pause'));
    });

    expect(onPauseMock).toHaveBeenCalledWith({ state: 'paused', count: 5 });
    expect(getByTestId('play')).not.toBeNull();
    expect(getByTestId('stop')).not.toBeNull();
  });

  it('stops the counter when the stop button is pressed', async () => {
    jest.useFakeTimers();
    const onStopMock = jest.fn();
    const { getByTestId, queryByTestId } = render(
      <Counter onStop={onStopMock} initialTimestamp={initialTimestamp} initialState="playing" />
    );

    act(() => {
      fireEvent.press(getByTestId('stop'));
      jest.advanceTimersByTime(1000);
    });

    expect(onStopMock).toHaveBeenCalledWith({ state: 'initial', count: 5 });
    expect(queryByTestId('play')).not.toBeNull();
    expect(queryByTestId('pause')).toBeNull();
    expect(queryByTestId('stop')).toBeNull();
  });
});
