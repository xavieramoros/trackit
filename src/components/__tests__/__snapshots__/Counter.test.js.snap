// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter  renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 40,
        }
      }
    >
      00:00:00
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "fontSize": 16,
            "marginTop": 4,
          }
        }
      >
        Play
      </Text>
    </View>
  </View>
</View>
`;
